src/types/createRouteOptions.ts

```diff
@@ -22,63 +22,68 @@ type WithHooks = {
   onAfterRouteLeave?: MaybeArray<AfterRouteHook>,
 }
 
-type WithSingleComponent = {
+export type WithHost<THost extends string | Host = string | Host> = {
   /**
-   * A Vue component, which can be either synchronous or asynchronous components.
+   * Host part of URL.
    */
-  component: Component,
+  host: THost,
 }
 
-type WithComponentRecord = {
-  /**
-   * Multiple components for named views, which can be either synchronous or asynchronous components.
-   */
-  components: Record<string, Component>,
+export function isWithHost(options: CreateRouteOptions): options is CreateRouteOptions & WithHost {
+  return 'host' in options && Boolean(options.host)
+}
+
+export type WithoutHost = {
+  host?: never,
 }
 
-type WithComponent = Partial<WithSingleComponent | WithComponentRecord>
+export type WithParent<TParent extends Route = Route> = {
+  parent: TParent,
+}
 
-export function isRouteWithParent(options: CreateRouteOptions): options is CreateRouteOptionsWithParent<Route> {
-  return 'parent' in options
+export function isWithParent<T extends Record<string, unknown>>(options: T): options is T & WithParent {
+  return 'parent' in options && Boolean(options.parent)
 }
 
-export function isRouteWithComponent(value: CreateRouteOptions): value is CreateRouteOptions & WithSingleComponent {
-  return 'component' in value
+export type WithoutParent = {
+  parent?: never,
+}
+
+export type WithComponent<
+  TComponent extends Component | undefined = Component | undefined
+> = {
+  /**
+   * A Vue component, which can be either synchronous or asynchronous components.
+   */
+  component?: TComponent,
```

Comment:
I'm surprised to see the `?` here. I think I personally preferred the `Partial<...>` route so `isWithComponent` would say `options is `WithComponent` but don't feel strongly if you think this is simpler

Reply:
I'd like the `With` types to reflect the use case. The `component` property is always optional when defining a route so I think its best to make it optional in the type itself rather than wrap it in `Partial` in the function definition. Plus the function types are a lot as is so reducing `Partial<WithComponent<TComponent>>` down to `WithComponent<TComponent>` makes it a little easier to read.

Reply:
@pleek91 Sounds good. Couldn't you define the type guard as `options is Required<WithComponent>`?